File: android/app/src/main/java/com/douxev/eggshell/data/voice/PitchDetector.kt
package com.douxev.eggshell.data.voice

/**
 * Monophonic pitch detector based on the YIN algorithm (de Cheveigné & Kawahara, 2002).
 *
 * Implemented inline rather than pulled in via TarsosDSP/JFreeChart because we
 * only need the single algorithm and don't want to ship ~700 KB of unrelated
 * DSP / charting code in the APK.
 *
 * Operates on a short window (≈ 0.05–0.1 s of audio at 44.1 kHz); the caller
 * is responsible for slicing the recording into windows and aggregating per-
 * window F0 estimates (see [estimateMedianHz]).
 *
 * Voiced-speech range is clamped to [70 Hz, 500 Hz]: trans HRT voice tracking
 * sits squarely in that band and rejecting extreme outliers gives a much
 * more stable per-clip headline number than reporting raw per-window F0.
 */
object PitchDetector {

    private const val DEFAULT_THRESHOLD = 0.15f
    private const val MIN_HZ = 70f
    private const val MAX_HZ = 500f
    private const val WINDOW_SIZE = 2048
    private const val HOP_SIZE = 1024
    private const val MIN_VOICED_WINDOWS = 6

    /**
     * Single-window YIN: returns the detected fundamental frequency in Hz, or
     * null if no clear pitch was found within the voiced-speech band.
     */
    fun detectHz(
        samples: FloatArray,
        sampleRate: Int,
        threshold: Float = DEFAULT_THRESHOLD,
    ): Float? {
        if (samples.size < 4) return null
        val tauMin = (sampleRate / MAX_HZ).toInt().coerceAtLeast(2)
        val tauMax = (sampleRate / MIN_HZ).toInt().coerceAtMost(samples.size / 2)
        if (tauMax <= tauMin) return null

        // Step 1: squared difference function d_t(τ) for τ in [1, tauMax].
        // We MUST compute it for every τ from 1, not just from tauMin, because
        // step 3 normalises by Σ d(j) for j=1..τ — skipping the low lags would
        // give an inflated cmnd in the τ ≈ tauMin region and bias detection
        // toward lower frequencies.
        val diff = FloatArray(tauMax + 1)
        for (tau in 1..tauMax) {
            var sum = 0f
            val limit = samples.size - tau
            var i = 0
            while (i < limit) {
                val d = samples[i] - samples[i + tau]
                sum += d * d
                i++
            }
            diff[tau] = sum
        }

        // Step 2: cumulative-mean-normalised difference d'_t(τ).
        // Definition: d'(τ) = d(τ) * τ / Σ_{j=1}^{τ} d(j), with d'(0) = 1.
        val cmnd = FloatArray(tauMax + 1)
        cmnd[0] = 1f
        var runningSum = 0f
        for (tau in 1..tauMax) {
            runningSum += diff[tau]
            cmnd[tau] = if (runningSum > 0f) diff[tau] * tau / runningSum else 1f
        }

        // Step 3: absolute threshold — first τ ≥ tauMin with d' < threshold.
        var tauEst = -1
        for (tau in tauMin..tauMax) {
            if (cmnd[tau] < threshold) {
                // Walk forward to the local minimum (YIN's "best local estimate").
                var t = tau
                while (t + 1 <= tauMax && cmnd[t + 1] < cmnd[t]) t++
                tauEst = t
                break
            }
        }
        if (tauEst < 0) return null

        // Step 4: parabolic interpolation around the dip for sub-sample resolution.
        val refinedTau = if (tauEst > tauMin && tauEst < tauMax) {
            val s0 = cmnd[tauEst - 1]
            val s1 = cmnd[tauEst]
            val s2 = cmnd[tauEst + 1]
            val denom = 2f * (2f * s1 - s0 - s2)
            if (denom != 0f) tauEst + (s2 - s0) / denom else tauEst.toFloat()
        } else {
            tauEst.toFloat()
        }
        if (refinedTau <= 0f) return null
        val f0 = sampleRate.toFloat() / refinedTau
        if (f0 !in MIN_HZ..MAX_HZ) return null
        return f0
    }

    /**
     * Walks the recording in overlapping windows and returns the median F0
     * across all voiced windows. Returns null when there aren't enough voiced
     * windows (silence / noise / non-vocal audio).
     *
     * The whole buffer is first passed through a one-pole high-pass filter at
     * 70 Hz: this kills AC hum, HVAC rumble, breath plosives and any DC bias
     * coming out of the AAC decoder. YIN is sensitive to low-frequency
     * energy because the cumulative-sum normalisation lets a long-period
     * rumble dominate the denominator and pull the detection toward an
     * octave-down estimate.
     */
    fun estimateMedianHz(samples: FloatArray, sampleRate: Int): Float? {
        if (samples.size < WINDOW_SIZE) return null
        val filtered = highPass(samples, sampleRate, cutoffHz = 70f)
        val window = FloatArray(WINDOW_SIZE)
        val voicedHz = ArrayList<Float>(filtered.size / HOP_SIZE)
        var offset = 0
        while (offset + WINDOW_SIZE <= filtered.size) {
            System.arraycopy(filtered, offset, window, 0, WINDOW_SIZE)
            // Skip near-silent frames so the median isn't dragged around by
            // background hiss that happens to alias to a pitch by accident.
            if (rms(window) >= 0.005f) {
                detectHz(window, sampleRate)?.let { voicedHz.add(it) }
            }
            offset += HOP_SIZE
        }
        if (voicedHz.size < MIN_VOICED_WINDOWS) return null
        voicedHz.sort()
        return voicedHz[voicedHz.size / 2]
    }

    /**
     * One-pole IIR high-pass filter.
     *
     * y[n] = α · (y[n-1] + x[n] - x[n-1])
     *
     * with α = RC / (RC + dt) and RC = 1 / (2π · fc). This is a 6 dB/octave
     * filter — gentle enough to leave voice fundamentals (70 Hz+) intact
     * while attenuating everything below by progressively more.
     */
    private fun highPass(samples: FloatArray, sampleRate: Int, cutoffHz: Float): FloatArray {
        val rc = 1f / (2f * Math.PI.toFloat() * cutoffHz)
        val dt = 1f / sampleRate
        val alpha = rc / (rc + dt)
        val out = FloatArray(samples.size)
        var prevX = samples[0]
        var prevY = 0f
        out[0] = 0f
        for (i in 1 until samples.size) {
            val y = alpha * (prevY + samples[i] - prevX)
            out[i] = y
            prevY = y
            prevX = samples[i]
        }
        return out
    }

    private fun rms(samples: FloatArray): Float {
        var sum = 0.0
        for (v in samples) sum += v * v
        return kotlin.math.sqrt(sum / samples.size).toFloat()
    }
}

File: apple/Eggshell/Sources/Voice/PitchDetector.swift
import Foundation
import AVFoundation

// ===========================================================================
// Monophonic fundamental-frequency (F0) estimator. Mirrors the Android
// PitchDetector + AudioDecoder pair:
//   1. decode the recorded .m4a to mono Float PCM via AVAudioFile →
//      AVAudioPCMBuffer (pure AVFoundation, no extra deps),
//   2. high-pass filter to kill DC / hum / rumble,
//   3. run the YIN algorithm (de Cheveigné & Kawahara, 2002) over
//      overlapping windows and take the median voiced F0.
//
// Implemented inline because we only need the single algorithm; the voiced
// band is clamped to 70–400 Hz, the plausible range for trans-HRT voice
// tracking — rejecting outliers gives a much steadier per-clip headline
// number than reporting raw per-window estimates.
// ===========================================================================

enum PitchDetector {

    private static let minHz: Float = 70
    private static let maxHz: Float = 400
    private static let threshold: Float = 0.15
    private static let windowSize = 2048
    private static let hopSize = 1024
    private static let minVoicedWindows = 6

    /// Decodes the clip at `url` and returns the median voiced F0 in Hz, or
    /// nil when the audio can't be decoded or no clear pitch is found.
    static func estimatePitch(url: URL) -> Int32? {
        guard let decoded = decodeToMonoFloats(url: url) else { return nil }
        guard let hz = estimateMedianHz(samples: decoded.samples, sampleRate: decoded.sampleRate) else {
            return nil
        }
        return Int32(hz.rounded())
    }

    // MARK: - Decoding (AVFoundation)

    private struct Decoded {
        let samples: [Float]
        let sampleRate: Float
    }

    /// Reads the audio file and converts it to mono Float PCM in [-1, 1].
    private static func decodeToMonoFloats(url: URL) -> Decoded? {
        guard let file = try? AVAudioFile(forReading: url) else { return nil }
        let inFormat = file.processingFormat
        let sampleRate = Float(inFormat.sampleRate)
        if sampleRate <= 0 { return nil }

        // Target: same sample rate, mono, deinterleaved Float32. Converting up
        // front means the YIN windows always see a single clean channel.
        guard let monoFormat = AVAudioFormat(
            commonFormat: .pcmFormatFloat32,
            sampleRate: inFormat.sampleRate,
            channels: 1,
            interleaved: false
        ) else { return nil }

        guard let converter = AVAudioConverter(from: inFormat, to: monoFormat) else { return nil }

        let frameCount = AVAudioFrameCount(file.length)
        if frameCount == 0 { return nil }

        guard let inputBuffer = AVAudioPCMBuffer(pcmFormat: inFormat, frameCapacity: frameCount) else {
            return nil
        }
        do {
            try file.read(into: inputBuffer)
        } catch {
            return nil
        }
        if inputBuffer.frameLength == 0 { return nil }

        // Converted capacity matches input frame count since the sample rate is
        // unchanged (mono downmix only).
        guard let outputBuffer = AVAudioPCMBuffer(pcmFormat: monoFormat, frameCapacity: inputBuffer.frameLength) else {
            return nil
        }

        var fed = false
        let inputBlock: AVAudioConverterInputBlock = { _, outStatus in
            if fed {
                outStatus.pointee = .endOfStream
                return nil
            }
            fed = true
            outStatus.pointee = .haveData
            return inputBuffer
        }

        var conversionError: NSError?
        let status = converter.convert(to: outputBuffer, error: &conversionError, withInputFrom: inputBlock)
        if status == .error || conversionError != nil { return nil }

        let n = Int(outputBuffer.frameLength)
        guard n > 0, let channelData = outputBuffer.floatChannelData else { return nil }
        let ptr = channelData[0]
        var samples = [Float](repeating: 0, count: n)
        for i in 0..<n { samples[i] = ptr[i] }
        return Decoded(samples: samples, sampleRate: sampleRate)
    }

    // MARK: - Median over windows

    /// Walks the buffer in overlapping windows and returns the median F0 of all
    /// voiced windows, or nil when too few windows are voiced.
    private static func estimateMedianHz(samples: [Float], sampleRate: Float) -> Float? {
        if samples.count < windowSize { return nil }
        let filtered = highPass(samples, sampleRate: sampleRate, cutoffHz: 70)
        var voiced: [Float] = []
        var window = [Float](repeating: 0, count: windowSize)
        var offset = 0
        while offset + windowSize <= filtered.count {
            for i in 0..<windowSize { window[i] = filtered[offset + i] }
            // Skip near-silent frames so background hiss doesn't drag the median.
            if rms(window) >= 0.005, let hz = detectHz(window, sampleRate: sampleRate) {
                voiced.append(hz)
            }
            offset += hopSize
        }
        if voiced.count < minVoicedWindows { return nil }
        voiced.sort()
        return voiced[voiced.count / 2]
    }

    // MARK: - Single-window YIN

    private static func detectHz(_ samples: [Float], sampleRate: Float) -> Float? {
        if samples.count < 4 { return nil }
        let tauMin = max(Int(sampleRate / maxHz), 2)
        let tauMax = min(Int(sampleRate / minHz), samples.count / 2)
        if tauMax <= tauMin { return nil }

        // Step 1: squared difference function d_t(τ) for τ in [1, tauMax].
        var diff = [Float](repeating: 0, count: tauMax + 1)
        for tau in 1...tauMax {
            var sum: Float = 0
            let limit = samples.count - tau
            var i = 0
            while i < limit {
                let d = samples[i] - samples[i + tau]
                sum += d * d
                i += 1
            }
            diff[tau] = sum
        }

        // Step 2: cumulative-mean-normalised difference d'_t(τ).
        var cmnd = [Float](repeating: 0, count: tauMax + 1)
        cmnd[0] = 1
        var runningSum: Float = 0
        for tau in 1...tauMax {
            runningSum += diff[tau]
            cmnd[tau] = runningSum > 0 ? diff[tau] * Float(tau) / runningSum : 1
        }

        // Step 3: absolute threshold — first τ ≥ tauMin with d' < threshold.
        var tauEst = -1
        var tau = tauMin
        while tau <= tauMax {
            if cmnd[tau] < threshold {
                var t = tau
                while t + 1 <= tauMax && cmnd[t + 1] < cmnd[t] { t += 1 }
                tauEst = t
                break
            }
            tau += 1
        }
        if tauEst < 0 { return nil }

        // Step 4: parabolic interpolation around the dip for sub-sample resolution.
        let refinedTau: Float
        if tauEst > tauMin && tauEst < tauMax {
            let s0 = cmnd[tauEst - 1]
            let s1 = cmnd[tauEst]
            let s2 = cmnd[tauEst + 1]
            let denom = 2 * (2 * s1 - s0 - s2)
            refinedTau = denom != 0 ? Float(tauEst) + (s2 - s0) / denom : Float(tauEst)
        } else {
            refinedTau = Float(tauEst)
        }
        if refinedTau <= 0 { return nil }
        let f0 = sampleRate / refinedTau
        if f0 < minHz || f0 > maxHz { return nil }
        return f0
    }

    // MARK: - DSP helpers

    /// One-pole IIR high-pass: y[n] = α · (y[n-1] + x[n] - x[n-1]).
    private static func highPass(_ samples: [Float], sampleRate: Float, cutoffHz: Float) -> [Float] {
        if samples.isEmpty { return samples }
        let rc = 1 / (2 * Float.pi * cutoffHz)
        let dt = 1 / sampleRate
        let alpha = rc / (rc + dt)
        var out = [Float](repeating: 0, count: samples.count)
        var prevX = samples[0]
        var prevY: Float = 0
        out[0] = 0
        for i in 1..<samples.count {
            let y = alpha * (prevY + samples[i] - prevX)
            out[i] = y
            prevY = y
            prevX = samples[i]
        }
        return out
    }

    private static func rms(_ samples: [Float]) -> Float {
        if samples.isEmpty { return 0 }
        var sum: Double = 0
        for v in samples { sum += Double(v) * Double(v) }
        return Float((sum / Double(samples.count)).squareRoot())
    }
}
